src/turbomind/kernels/decoder_masked_multihead_attention/decoder_masked_multihead_attention_template.cuh

```diff
@@ -1461,47 +1416,23 @@ __global__ void masked_multihead_attention_kernel(Multihead_attention_params<T>
         // The position of the thread in that 16B chunk.
         int ci = tidx % QK_VECS_IN_16B * QK_VEC_SIZE;
 
-        if (handle_kv && group_leader) {
+        if (group_leader) {
             // Trigger the stores to global memory.
             if (Dh == Dh_MAX || co < Dh / QK_ELTS_IN_16B) {
-                if (!params.k_cache_per_sample) {
-                    // Two chunks are separated by L * x elements. A thread write QK_VEC_SIZE elements.
-                    int offset = bhi * params.memory_max_len * Dh + co * params.memory_max_len * QK_ELTS_IN_16B
-                                 + tlength_circ * QK_ELTS_IN_16B + ci;
 
-                    if (not QUANT_POLICY) {
-                        *reinterpret_cast<Qk_vec_m*>(&params.k_cache[offset]) = vec_conversion<Qk_vec_m, Qk_vec_k>(k);
-                    }
-                    else if (QUANT_POLICY == 4) {
-                        using Packed_Int8_t  = typename packed_type<int8_t, num_elems<Qk_vec_k>::value>::type;
-                        Packed_Int8_t k_int8 = quant(k, k_scale, k_zp);
+                int offset = params.kv_cache_per_sample_offset + kvhi * params.memory_max_len * Dh + tlength_circ * Dh
+                             + co * QK_ELTS_IN_16B + ci;
 
-                        int8_t* dst_ptr                                     = reinterpret_cast<int8_t*>(params.k_cache);
-                        *reinterpret_cast<Packed_Int8_t*>(&dst_ptr[offset]) = k_int8;
-                    }
+                if (not QUANT_POLICY) {
```

Comment:
May merge latest main branch. "if (not ...)" isn't supported on Windows platform

Reply:
Please update "News" section and "Supported Models" section in both README